### PR TITLE
feat: implement booking & item modules with validation, comments, tests

### PR DESCRIPTION
**Pull request by a-buianova**
Implemented sprint requirements:

- Booking:
  • Added Booking entity, repository, service and controller.
  • Implemented booking creation with validations (time window, availability, owner cannot book own item).
  • Added approval flow (WAITING → APPROVED/REJECTED) restricted to item owners.
  • Implemented retrieval for booker/owner and paginated listing with state filtering (ALL, CURRENT, PAST, FUTURE, WAITING, REJECTED).
  • Mapper handles conversion between Instant (DB) and LocalDateTime (API DTOs).
  • Unit tests for mapper and service cover edge cases (invalid intervals, forbidden access, nulls).

- Item:
  • Added item CRUD (create, patch, search, list by owner, details).
  • Implemented last/next bookings for owner in item details.
  • Added comments: only users with completed APPROVED bookings can comment.
  • Mappers for items and comments, with DTO layer aligned to API spec.

- Common:
  • Global exception handler with consistent error responses (400/403/404/409/500).
  • Custom exceptions for domain rules (BadRequest, NotFound, Conflict, Forbidden).
  • Unified @CurrentUserId resolver for headers.
  • Database schema updated (constraints, indexes, FK relations).

- Tests & Stability:
  • Fixed Postman and unit tests, adjusted time zone handling (system default vs UTC).
  • Achieved full compatibility with sprint Postman collection.
  • Cleaned up code style, Javadoc in English, consistent package structure.

### CODE REVIEW

File: src/main/java/ru/practicum/shareit/booking/controller/BookingController.java [hunk 1]
```diff
@@ -1,28 +1,69 @@
...
+    public BookingResponse create(@CurrentUserId Long userId,
+                                  @RequestBody @Valid BookingCreateDto dto) {
+        return service.create(userId, dto);
+    }
+
+    /** PATCH /bookings/{id}?approved={true|false} — owner approves/rejects. */
+    @PatchMapping("/{bookingId}")
+    public BookingResponse approve(@CurrentUserId Long ownerId,
+                                   @PathVariable Long bookingId,
+                                   @RequestParam("approved") boolean approved) {
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/booking/controller/BookingController.java]: Если имя параметра запроса и аргумента совпадают, то можно не указывать имя

File: src/main/java/ru/practicum/shareit/booking/repo/BookingRepository.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            """)
+    List<Booking> findFutureByOwner(Long ownerId, Instant after, Pageable pageable);
+
+    @Query("""
+            select b from Booking b
+            where b.item.owner.id = :ownerId
+              and b.status = :status
+            order by b.start desc
+            """)
+    List<Booking> findByOwnerAndStatus(Long ownerId, BookingStatus status, Pageable pageable);
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/booking/repo/BookingRepository.java]: Запросы достаточно простые и можно обойтись без `@Query`

File: src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    @Transactional
+    public BookingResponse create(Long userId, BookingCreateDto dto) {
+        User booker = userRepo.findById(userId)
+                .orElseThrow(() -> new NotFoundException("user not found"));
+        Item item = itemRepo.findById(dto.itemId())
+                .orElseThrow(() -> new NotFoundException("item not found"));
+
+        LocalDateTime start = dto.start();
+        LocalDateTime end   = dto.end();
+        if (start == null || end == null || !start.isBefore(end)) {
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java]: На null же валидируется в дто

File: src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            throw new BadRequestException("invalid time window");
+        }
+        if (!item.isAvailable()) {
+            throw new BadRequestException("item is not available");
+        }
+        if (item.getOwner() != null && item.getOwner().getId().equals(userId)) {
+            throw new NotFoundException("owner cannot book own item");
+        }
+
+        Booking saved = bookingRepo.save(BookingMapper.toEntity(dto, item, booker));
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:65]: Еще нужно перед созданием брони проверить, что она не пересекается по времени с уже созданными

File: src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        var page = PageRequest.of(from / size, size);
+        Instant now = Instant.now();
+
+        List<Booking> data = switch (state) {
+            case ALL      -> bookingRepo.findByBooker_IdOrderByStartDesc(userId, page);
+            case CURRENT  -> bookingRepo.findByBooker_IdAndStartBeforeAndEndAfterOrderByStartDesc(userId, now, now, page);
+            case PAST     -> bookingRepo.findByBooker_IdAndEndBeforeOrderByStartDesc(userId, now, page);
+            case FUTURE   -> bookingRepo.findByBooker_IdAndStartAfterOrderByStartDesc(userId, now, page);
+            case WAITING  -> bookingRepo.findByBooker_IdAndStatusOrderByStartDesc(userId, BookingStatus.WAITING, page);
+            case REJECTED -> bookingRepo.findByBooker_IdAndStatusOrderByStartDesc(userId, BookingStatus.REJECTED, page);
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java]: Можно объединить в один case, тут запросы одинаковые, отличается только статус

File: src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        var page = PageRequest.of(from / size, size);
+        Instant now = Instant.now();
+
+        List<Booking> data = switch (state) {
+            case ALL      -> bookingRepo.findAllByOwner(ownerId, page);
+            case CURRENT  -> bookingRepo.findCurrentByOwner(ownerId, now, page);
+            case PAST     -> bookingRepo.findPastByOwner(ownerId, now, page);
+            case FUTURE   -> bookingRepo.findFutureByOwner(ownerId, now, page);
+            case WAITING  -> bookingRepo.findByOwnerAndStatus(ownerId, BookingStatus.WAITING, page);
+            case REJECTED -> bookingRepo.findByOwnerAndStatus(ownerId, BookingStatus.REJECTED, page);
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java]: И тут

File: src/main/java/ru/practicum/shareit/common/error/GlobalExceptionHandler.java [hunk 7]
```diff
@@ -115,6 +115,14 @@ public ErrorResponse handleUnreadable(HttpMessageNotReadableException ex, HttpSe
         return body(HttpStatus.BAD_REQUEST, message, req);
     }
 
+    @ExceptionHandler(IllegalArgumentException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleIllegalArgument(IllegalArgumentException ex, HttpServletRequest req) {
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/common/error/GlobalExceptionHandler.java]: Это слишком обширное исключение, которое может быть выброшено в разных случаях, даже которых мы не ожидаем. Нужно в обработчиках ловить только то, что мы допускаем в логике осознанно

File: src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java [hunk 8]
```diff
@@ -2,67 +2,139 @@
...
+            Long itemId = i.getId();
+            Booking last = bookingRepo
+                    .findTopByItem_IdAndStatusAndStartBeforeOrderByStartDesc(itemId, BookingStatus.APPROVED, now)
+                    .orElse(null);
+            Booking next = bookingRepo
+                    .findTopByItem_IdAndStatusAndStartAfterOrderByStartAsc(itemId, BookingStatus.APPROVED, now)
+                    .orElse(null);
+            List<Comment> comments = commentsByItem.getOrDefault(itemId, List.of());
+            out.add(mapper.toDetails(i, last, next, comments));
+        }
```
Comment (Ivanov-s-tmn) [inline src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java]: Запросы в цикле (проблема N + 1)

File: src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java
Comment (Ivanov-s-tmn) [inline src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:1]: Так как это юнит тесты, то нужно мокать сервисный слой, к которому обращается контроллер. Почему используются репозитории, если контроллер обращается к сервису?

File: src/test/java/ru/practicum/shareit/booking/BookingRepositoryIT.java
Comment (Ivanov-s-tmn) [inline src/test/java/ru/practicum/shareit/booking/BookingRepositoryIT.java:1]: Если тесты интеграционные, то нужно в названии класса тестов это указать

File: src/main/java/ru/practicum/shareit/booking/controller/BookingController.java (hunk 1 above)
Comment (a-buianova) [reply src/main/java/ru/practicum/shareit/booking/controller/BookingController.java]: Согласна: имя можно опустить, оставила его из явности в доках для Postman

File: src/main/java/ru/practicum/shareit/booking/repo/BookingRepository.java (hunk 2 above)
Comment (a-buianova) [reply src/main/java/ru/practicum/shareit/booking/repo/BookingRepository.java]: Спасибо! Простые запросы на списки заменила на derived-методы. @Query оставила только там, где он предотвращает N+1 (batch last/next по множеству items) и где выражаемость derived-методов слабая (проверка пересечений). Надеюсь верно =)

**Comment by a-buianova** [comment]
Спасибо большое за ревью @Ivanov-s-tmn .  Все вроде замечания поправила, старалась учесть и в других тестах тоже. Полезные замечания, помогают смотреть на код внимательнее.

File: src/test/java/ru/practicum/shareit/booking/BookingControllerIT.java
Comment (Ivanov-s-tmn) [inline src/test/java/ru/practicum/shareit/booking/BookingControllerIT.java:1]: Тесты зависят от порядка выполнения

Это антитипаттерн: каждый тест должен быть независимым
Если изменить порядок или пропустить тест — всё сломается
Усложняет параллельное выполнение

File: src/test/java/ru/practicum/shareit/booking/BookingControllerIT.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+        mvc.perform(get("/bookings/owner").param("state", "ALL").header(USER_HEADER, ownerId))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$", hasSize(greaterThanOrEqualTo(1))));
+    }
+
+    @Test
+    @Order(5)
+    @DisplayName("Ошибки: отсутствие заголовка и плохой state → 400")
+    void errors_400() throws Exception {
```
Comment (Ivanov-s-tmn) [inline src/test/java/ru/practicum/shareit/booking/BookingControllerIT.java]: больше одного сценария в тесте проверять не стоит, 1 сценарий - 1 тест
